### PR TITLE
File resuming for utox

### PR DESCRIPTION
**Pull request by GrayHatter**

### CODE REVIEW

**Comment by GrayHatter** [comment]
@irungentoo transfers work both ways, and resume on both sides.

Let me know if I missed anything!

File: file_transfers.c [hunk 1]
```diff
@@ -3,9 +3,13 @@
 //static FILE_TRANSFER *file_t[256], **file_tend = file_t;
 static FILE_TRANSFER outgoing_transfer[MAX_NUM_FRIENDS][MAX_FILE_TRANSFERS];
 static FILE_TRANSFER incoming_transfer[MAX_NUM_FRIENDS][MAX_FILE_TRANSFERS];
+static BROKEN_TRANSFER broken_list[33] = {{0}}; /* TODO De-hardcode this. 33 because we can't use 0.
```
Comment (irungentoo) [inline file_transfers.c]: Make that 33 a define or something.

File: file_transfers.c [hunk 2]
```diff
@@ -59,75 +62,208 @@ static void calculate_speed(FILE_TRANSFER *file)
...
-        } else if(file->status == FILE_TRANSFER_STATUS_PAUSED_THEM) {
-            // Do nothing;
+static int utox_file_alloc_resume(Tox *tox, FILE_TRANSFER *file){
+    TOX_ERR_FILE_GET error;
+    int i = 0;
+    int resume = 0;
+    resume = tox_file_get_file_id(tox, file->friend_number, file->file_number, file->file_id, &error);
+    if (resume){
+        file->resume = 0;
+        for(i = 1; i <= 33; i++){
```
Comment (irungentoo) [inline file_transfers.c]: index 33 is out of bounds.

File: file_transfers.c [hunk 3]
```diff
@@ -620,3 +1017,42 @@ void utox_set_callbacks_for_transfer(Tox *tox){
...
+    BROKEN_TRANSFER *saved = calloc(33, sizeof(BROKEN_TRANSFER));
+    saved = file_raw((char*)path, &size_read);
+    if(saved && size_read == (sizeof(BROKEN_TRANSFER) * 33)){
+        memcpy(broken_list, saved, (sizeof(BROKEN_TRANSFER) * 33));
+    } else {
+        debug("SaveFile:\tUnable to load uTox Save file for File Transfers\n");
+    }
+    free(saved);
+
+    for(int i = 0 ; i <= 33 ; i++){
```
Comment (irungentoo) [inline file_transfers.c]: index 33 is out of bounds.

**Comment by irungentoo** [comment]
Clicking the accept file transfer button starts the transfer even when a final path isn't chosen. If the transfer finishes before a path is picked, uTox crashes.

**Comment by irungentoo** [comment]
resumed files get corrupted.

File: file_transfers.c [hunk 4]
```diff
@@ -59,75 +61,208 @@ static void calculate_speed(FILE_TRANSFER *file)
...
+            if(!broken_list[i].used){
+                // TODO set an expire time for this.
+                broken_list[i].used = 1;
+                broken_list[i].incoming = file->incoming;
+
+                broken_list[i].friend_number = file->friend_number;
+                broken_list[i].file_number = file->file_number;
+                memcpy(broken_list[i].file_id, file->file_id, sizeof(uint8_t) * TOX_FILE_ID_LENGTH);
+                if(file->path){
+                    memcpy(broken_list[i].file_path, file->path, UTOX_FILE_NAME_LENGTH);
```
Comment (irungentoo) [inline file_transfers.c]: This memcpy causes a segfault.

**Comment by GrayHatter** [comment]
@irungentoo I've tested it every way I can possibly think of; win/lin threw in random data, asan'd every action I could try starting, pausing, breaking, restarting, rebooting.

I'd put money on this being ready to merge into master

**Pull request merged**